### ISSUE TITLE
Airplay streaming to HomePod randomly stops

### ISSUE THREAD

**Issue by maxclaey** [ejurgensen/forked-daapd#521]
I'm currently experimenting with running forked-daapd on a Raspberry Pi 3 to be able to play Internet radio on my HomePod.
Initially, this works like a charm. However, I notice that the connection to the HomePod randomly drops every couple of minutes, causing the playout to stop.

Any ideas on why this problem occurs, or any suggestions on how to solve this issue?

Below you can find a log extract from the moment the playout stops:
```
[2018-04-08 10:31:21] [DEBUG]   player: Input buffer has 355400 bytes
[2018-04-08 10:31:26] [DEBUG]   player: Action is metadata_get
[2018-04-08 10:31:26] [ WARN]   player: Output delay detected: player is 2 ticks behind, catching up
[2018-04-08 10:31:31] [DEBUG]   player: Action is metadata_get
[2018-04-08 10:31:31] [DEBUG]   player: Input buffer has 356584 bytes
[2018-04-08 10:31:36] [DEBUG]   player: Action is metadata_get
[2018-04-08 10:31:41] [  LOG]     raop: Device 'Woonkamer' closed RTSP connection
[2018-04-08 10:31:41] [DEBUG]     raop: Cleaning up failed session (deferred) on device 'Woonkamer'
[2018-04-08 10:31:41] [DEBUG]   player: Callback from AirPlay to device_streaming_cb
[2018-04-08 10:31:41] [  LOG]   player: The AirPlay device 'Woonkamer' FAILED
[2018-04-08 10:31:41] [DEBUG]     raop: Building TEARDOWN for 'Woonkamer'
[2018-04-08 10:31:41] [DEBUG]   player: Pause called, stopping playback loop
[2018-04-08 10:31:41] [DEBUG]   player: Playback loop stopped (break is 1, ret 0)
[2018-04-08 10:31:41] [DEBUG]   player: Flushing 356268 bytes with flags 8
[2018-04-08 10:31:41] [DEBUG]   player: Action is stop
[2018-04-08 10:31:41] [DEBUG]       db: Running query 'BEGIN TRANSACTION;'
[2018-04-08 10:31:41] [DEBUG]       db: Running query 'SELECT value FROM admin a WHERE a.key = 'queue_version';'
[2018-04-08 10:31:41] [DEBUG]       db: Running query 'DELETE FROM queue where id <> 0;'
[2018-04-08 10:31:41] [DEBUG]       db: Running query 'INSERT OR REPLACE INTO admin (key, value) VALUES ('queue_version', '340');'
[2018-04-08 10:31:41] [DEBUG]       db: Running query 'END TRANSACTION;'
[2018-04-08 10:31:41] [DEBUG]      mpd: Asynchronous listener callback called with event type 2.
[2018-04-08 10:31:41] [DEBUG]      mpd: Asynchronous listener callback called with event type 1.
[2018-04-08 10:31:41] [DEBUG]      mpd: Notify clients waiting for idle results: 2
[2018-04-08 10:31:41] [DEBUG]      mpd: Notify clients waiting for idle results: 1

**Comment by ejurgensen** [comment]
Is it at a precise time playback stops? Or is there randomness to it?

For ATV had to implement a little hack to counter similar behavior, maybe the same is required for HomePod. Don't have one to test, though.

Are you able to build forked-daapd yourself if I give you some instuctions on how to enable the hack? Otherwise I can probably help you with a binary.

**Comment by maxclaey** [comment]
Thanks for your swift response! There's for sure some randomness. Sometimes it happens after a couple of seconds, sometimes its only after playing for 15-20 min. I am currently working with forked-daapd build from git

**Comment by ejurgensen** [comment]
Ok, if there is randomness then it is likely not the same problem as the ATV. The ATV disconnects were after a fixed time. Since you build from git you can test yourself easily, just set rs->keep_alive to 1 here: https://github.com/ejurgensen/forked-daapd/blob/master/src/outputs/raop.c#L1991 (I expect a HomePod gets detected as an Other type of device, but you can verify that from forked-daapd's log).

Another possibility is an underrun, i.e. the HomePod running out of audio to play. But that doesn't seem so likely if it sometimes happens after a few seconds.

Do you always see the message "Output delay detected: player is XXXX ticks behind, catching up" right before it happens? Output delay means that forked-daapd has trouble sending data to the device. It could be network related. However, 2 ticks (like in your log) should be insignificant, so that also seems a bit unlikely.

**Comment by maxclaey** [comment]
Ok, great, thanks for the suggestions. I only ran once with logging on debug to be able to get more detailed output, so I don't know if the message is always there. I will do some more testing and try with the ATV fix as well. I probably won't be able to test today, but I'll keep you posted for sure.

Thanks again!

**Comment by maxclaey** [comment]
I checked with the fix you propose and things for sure improved. Occasionally I still experience stopping streams, but it looks like they are network related, based on these logs:
```
[2018-04-09 16:56:44] [  LOG]   player: Source is not providing sufficient data, temporarily suspending playback (deficit=188)
[2018-04-09 16:56:55] [  LOG]    xcode: Timeout while reading source (connection problem?)
[2018-04-09 16:56:55] [ WARN]   ffmpeg: invalid new backstep -1
[2018-04-09 16:56:55] [ WARN]    xcode: Could not read frame: Immediate exit requested
[2018-04-09 16:57:09] [ WARN]    event: evrtsp_read: evbuffer_read: Connection reset by peer
[2018-04-09 16:57:09] [  LOG]   player: The AirPlay device 'Woonkamer' FAILED
[2018-04-09 16:57:19] [  LOG]     mdns: Avahi Resolver failure: service '18EE6922D238@Woonkamer Apple TV' type '_raop._tcp' proto 0: Timeout reached
[2018-04-09 16:57:19] [  LOG]     mdns: Avahi Resolver failure: service 'D4A33D62D336@Woonkamer' type '_raop._tcp' proto 0: Timeout reached
[2018-04-09 16:57:19] [ INFO]   player: Removing AirPlay device 'Woonkamer Apple TV'; stopped advertising
[2018-04-09 16:57:19] [ INFO]   player: Removing AirPlay device 'Woonkamer'; stopped advertising
[2018-04-09 16:57:21] [  LOG]     mdns: Avahi Resolver failure: service 'B8782E5708CF@Slaapkamer Apple TV' type '_raop._tcp' proto 0: Timeout reached
[2018-04-09 16:57:21] [ INFO]   player: Removing AirPlay device 'Slaapkamer Apple TV'; stopped advertising

**Comment by ejurgensen** [comment]
The log indicates that you are listening to an internet stream? For testing purposes it would be good if it was local audio (mp3 files or something), otherwise it will be difficult to isolate if the interruptions are caused by the internet stream, or caused by connection issues between forked-daapd and the HomePod.

**Comment by maxclaey** [comment]
Indeed, I am. I am using forked-daapd to have internet radio capabilities on my HomePod (a feature I'm missing there), so I only listen to internet streams through it. If it would be useful to you, I can try the with local audio, but I'm quite convinced that the issues I still see right now are connectivity-related

**Comment by maxclaey** [comment]
After testing this a bit more I can confirm that the issue is solved by setting keep_alive to 1 for the HomePod. However, I just noticed another, rather annoying issue with streaming to HomePod :). The volume in the forked-daapd player is set to 15. When I change the volume on HomePod directly by saying "Hey Siri, set the volume to 25%" the volume increases, but falls back to 15% a couple of seconds later. Any workaround for this behaviour? Many thanks!

**Comment by ejurgensen** [comment]
Well, the volume adjustment is actually caused by the keep_alive setting. The way forked-daapd keeps the device alive is by sending a volume command every minute. Of course that doesn't work so well in your scenario.

I'll need to figure out some other way of keeping the connection alive. Could be a bit difficult, also due to me having neither an ATV4 nor a Homepod.

**Comment by maxclaey** [comment]
Hmmm ok, I see.. I am not at all familiar with what is possible with the airplay protocol, but if there would be a way to fetch the current volume level prior to setting it, the problem would be solved I guess?

**Comment by ejurgensen** [comment]
Yes, I think that could work. Maybe just requesting the volume at regular intervals is enough to keep the connection alive, and that would have the added advantage of being able to show the actual, current volume setting in forked-daapd. Will look at that asap!

**Comment by maxclaey** [comment]
Great! If I can be of any help, please let me know!

**Comment by ejurgensen** [comment]
Yes, you certainly can - it would be great if you could build and test from this branch:
https://github.com/ejurgensen/forked-daapd/tree/keep_alive1

In that branch, I have made a change so that forked-daapd gets the volume (every 30 sec) instead of setting it. I can't test if that will still have the effect of keeping the connection alive, since I don't have an ATV4 or Homepod. So I hope you can give it a try. It would be best if you tested with a local media file.

**Comment by maxclaey** [comment]
I am currently testing the `keep_alive1` branch (+ setting keep_alive to 1 on line 2041) and things seem to be fine! Music is streaming for 3.5 hours straight now, without interruptions. Changing the volume on the HomePod itself is now maintained as well! Only thing I noticed is that even though I change the volume on the HomePod, the volume reported by the json api when requesting the player state does not change. However, for me this for sure is not an issue. I'm really happy with this update!

**Comment by ejurgensen** [comment]
That's good to hear. I didn't yet make the part that sets the volume internally in forked-daapd. Volume is not quite so simple as one would imagine, due to forked-daapd supporting multiple speakers. There are master and device volumes, and the latter can both be relative to the master and absolute. So it will take some work.

**Comment by ejurgensen** [comment]
I've merged the change into master now (commit d9a67b6 and acd5bd3). It is excluding the part about reading the volume into forked-daapd, I will save that for later.

**Comment by ejurgensen** [comment]
@maxclaey, I'm planning on making a change so that forked-daapd will make so-called OPTIONS requests instead of polling for volume. Polling for volume causes problems on old Apple Express.

I've made the change in this branch: https://github.com/ejurgensen/forked-daapd/tree/fix_keep_alive1

Any chance you could test it? Don't want to break the HomePod, of course.

**Comment by maxclaey** [comment]
@ejurgensen Sure, I will try it out ASAP and let you know what it gives on HomePod!

**Comment by maxclaey** [comment]
@ejurgensen I just tested your `fix_keep_alive1` branch and can confirm that it doesn't break HomePod support. Music is streaming for 40min straight now, without issues

**Comment by ejurgensen** [comment]
Great, thanks for testing. I have merged the change now.

**Comment by ejurgensen** [comment]
FYI I have made some changes (commit b6e1269) that could mean that volume changes from the Homepod are received and understood by forked-daapd. Meaning they should also be reflected via the json api, and via clients like Apple Remote.

The changes may also mean that you control playback (pause/play/next/previous) via the Homepod.

It would be great if you could test and let me know if it works.

**Comment by maxclaey** [comment]
I just tested it and you are right: the volume changes one Homepod are reflected just fine!

**Comment by retrography** [comment]
Has this been merged in the master already? I am on 26.5 and I seem to have the same issue with my HomePod.

**Comment by ejurgensen** [comment]
Yes, the above was merged quite a while ago, and it is included in 26.5. Maybe your issue is the same as #496 which I have not been able to solve?

**Comment by retrography** [comment]
Can't be that one, because I can immediately re-enable the speakers. If I am multicasting, I can even re-enable the speakers to continue listening to the same song (which was meanwhile being streamed to the other speakers). 

I have compile and installed the new version over debian's original distribution, hoping that "make install" overwrites the old version. Issuing the version command gives me 26.5. Is there any reason to doubt that version report (for instance, is it possible that some essential modules were not updated)?

**Comment by ejurgensen** [comment]
The version report is correct, but depending on what arguments you gave to "configure" you could have a debian binary of forked-daapd in /usr/sbin and your own build in say /usr/local/sbin. And it might be the Debian bin that gets started on boot. So if building yourself I would recommend removing the Debian version.

It is of course also possible that you have encountered some other issue with forked-daapd and HomePod. If that turns out to be the case please open a new issue with relevant logging info (with the log set to debug level in the config file).

**Comment by retrography** [comment]
Uninstalled the deb package and reinstalled from the source code. Same problem. Will file a bug report along with the log file.
